### ISSUE TITLE
Bug: [array-type] rule mistaking a generic type with the actual type because it has the same name (shadowing).

### ISSUE THREAD

**Issue by miguel-leon** [typescript-eslint/typescript-eslint#11287]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&fileType=.ts&code=GYVwdgxgLglg9mABMMAeAggJ0wQwJ6ICmAHlIWACYDOimhOFCANgeANZhwDuYA2gLoA%2BABQ5s%2BAFyIsuPAEpEAbwCwAKES1CUEJiRjZAbjUBfI6qA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQ2mncNqNPRREXAPbRI4MAF8QUoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function fn<Array extends readonly unknown[]>(array: Array) {
  return array;
};
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/array-type": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No errors.

### Actual Result

Error on the generic type parameter.

**Comment by bradzacher** [comment]
This is expected - the rule, like the majority of our rules, does syntactic checks.

The reason we don't do anything more expensive or heavier is because it's widely considered bad practice to shadow well-known global names like this.

Many eslint rules - not just rules about types - work this way and make similar assumptions.

**Comment by miguel-leon** [comment]
So, if you shouldn't use the built-in type, and you also shouldn't shadow the name, not even for a parameter, the rule is effectively just "let's code with fewer words". Nice rule, it should be renamed to `banned-identifiers` or something.
